### ISSUE TITLE
Multiple optimizers support

### ISSUE THREAD

**Issue by galatolofederico** [MillionIntegrals/vel#53]
Some architectures like GANs use two or more optimizers and it would be nice to add this behavior to vel.
vel right now relies on the function `calculate_gradient` to compute the loss backward. What do you think if instead the `calculate_gradient` will return a list of loss tensors (and metrics) and the Trainer will also have a list of optimizers and the backward() will be called by the Trainer in the `train_batch` function.

I am thinking about something like
```
for loss, optimizer in zip(self.model.calculate_gradient(), self.model.optimizers):
   optimizer.zero_grad()
   loss.backward()
   #clip gradients
   optimizer.step()
```

The `calculate_gradient` can be also retro compatible by checking the return type (dict or list).

What do you think? if you are ok with it i can work on it :smiley:

**Comment by MillionIntegrals** [comment]
Hi!

Thank you for using the software! Funny thing, I was thinking about very same thing recently, as that's one of the least extensible parts of the framework and I was also thinking of adding some GAN support. 

I still need to figure out what the best API will be for this usecase, but I can come back to you around the end of the week with some ideas.

**Comment by galatolofederico** [comment]
Lets keep in touch this weekend!
In the mean time what do you think about something like this:

Add a class to `vel.api.model` named `MultipleOptimizersModel` that extends `GradientModel`

`MultipleOptimizersModel` add this functions to the interface:
-  `named_parameters()` 
-  `named_losses()`

`named_parameters()` returns `dict` with this structure:
```
{
 "parameters_group_name_1" : [list of parameters 1],
 "parameters_group_name_2" : [list of parameters 2],
...
}
```
`named_losses` returns a `dict` with this structure:
```
{
 "parameters_group_name_1" : loss_tensor_for_group_name_1,
 "parameters_group_name_2" : loss_tensor_for_group_name_2
...
}
```

In the yaml you have to configure multiple optimizers like this:
```
optimizer:
  parameters_group_name1:
    name: vel.optimizer.sgd
    lr: 0.1
  parameters_group_name2:
    name: vel.optimizer.adam
    lr: 0.1
```

`MultipleOptimizersModel` implements `calculate_gradient()` with a for loop that matches the keys in the above dictionaries and computes the `optimizer.zero_grad()` `loss.backward()` and `optimizer.step()` for each one.

**Comment by MillionIntegrals** [comment]
This is a good idea, although there is still a question whether these losses can be returned together or maybe they have to be computed sequentially and optimized separately. Maybe it's better to just have model implement something like an `optimize` method, taking an optimizer (or a list of optimizers) and let it do whatever it wants... 

These are definitely some ideas to work with here, there are still some other things I'd like to do this week, but I'll try to implement something like that and a simple GAN example late next week.

**Comment by MillionIntegrals** [comment]
I've just pushed a pretty heavy refactor with super basic support for multiple optimizers and a very simple GAN training example (that I'm not sure it even works yet). I'll be refining it over few following weeks. Let me know if you think this solution is reasonable.

**Comment by galatolofederico** [comment]
Hey sorry for the delay, I had a crazy work week :smiley: 
I've checked out the solution, the use of just an `optimize` function is very clean and easy to extend. Nice work!

**Comment by MillionIntegrals** [comment]
I hope you'll like it. I'm sorry for the heavy refactoring recently, but there's quite a bit of trade-offs in trying to make all these things somehow compatible.

Nevertheless I think I pushed the last one large-ish update today and the API of 0.4 should be getting somewhat stable. I'll still do some clean-ups here and there, but mostly will focus on implementing a few new models, benchmarking the old ones to make sure they still work well and I target to release new version sometime later this year before NeurIPS.
